### ISSUE TITLE
SqlException: Must declare the table variable @p0

### ISSUE THREAD

**Issue by vovanthinh2205** [mishael-o/Dapper.SimpleSqlBuilder#87]
### What happened?

I was really impressed with Dapper.SimpleSqlBuilder.so I decided to use this library for my project.but when I install Dapper.SimpleSqlBuilder and wrote my sample code with your guide on web https://mishael-o.github.io/Dapper.SimpleSqlBuilder/.There is error SqlException: Must declare the table variable @p0.could you explain why I am error that. I open a new project with .net 8 and a database SQL server.
I use libraries such as Dapper" Version="2.1.35",Dapper.SimpleSqlBuilder Version="3.7.4",Dapper.SimpleSqlBuilder.DependencyInjection Version="3.7.4" 

![Capture](https://github.com/mishael-o/Dapper.SimpleSqlBuilder/assets/163615264/29d6604a-4db8-41f2-80cd-11ca70700170)

### Reproduction steps

1.
2.
3.
...

### Version (bug)

3.7.4

### Version (working)

3.7.4

### Operating System

Windows

### Pull Request

- [X] I would like to do a Pull Request

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by vovanthinh2205** [comment]
![Uploading Capture.PNG…]()

**Comment by vovanthinh2205** [comment]
![Capture](https://github.com/mishael-o/Dapper.SimpleSqlBuilder/assets/163615264/3a51f5eb-26ee-4d44-9c53-84f47c6d5cd1)

**Comment by mishael-o** [comment]
Hi, @vovanthinh2205, thank you, I am glad the library is useful to you.

As for the issue you mentioned, you can't parameterize table names, they are part of the prepared statement and the MSSQL database (and others) do not allow this. You should just put the table name in the SQL query `SELECT * FROM Product`.

However, if the table name is dynamic then you can use the `:raw` (https://mishael-o.github.io/Dapper.SimpleSqlBuilder/docs/advanced-features/raw-values.html) string format to tell the library not to parameterize the table name but instead use the value. See an example below.

> **Note: The `:raw` string format can lead to SQL injection if you do not sanitize the value, so make sure you do**

```csharp
using (var connection = new SqlConnection(connectionString))
{
    var builder = SimpleBuilder.Create($"SELECT * FROM {table:raw}");
    var list = connection.Query<object>(builder.Sql, builder.Parameters);
}
```

**Comment by vovanthinh2205** [comment]
Thank you so much! I fixed that error.
I had a new problem when I used Dependency Injection with Dapper.SimpleSqlBuilder.DependencyInjection Version="3.7.4"
and my config of your sample on your web. but it notifies to a new error like  serviceLifeTime not  exist in the current context.

![image](https://github.com/mishael-o/Dapper.SimpleSqlBuilder/assets/163615264/d7abe37b-0770-4bf8-8198-fd2bc01fb822)

**Comment by mishael-o** [comment]
> Thank you so much! I fixed that error. I had a new problem when I used Dependency Injection with Dapper.SimpleSqlBuilder.DependencyInjection Version="3.7.4" and my config of your sample on your web. but it notifies to a new error like serviceLifeTime not exist in the current context.

Oh, that's a typo in the docs, it's meant to be `:` and not `=`. I will fix that (see the correct code below). Thanks.

```chsarp
builder.Services.AddSimpleSqlBuilder(
    serviceLifetime: ServiceLifetime.Singleton);
```

Also note if you are using the Depenjection Injection package, make sure you inject the `ISimpleBuilder` and use that to create the builder instances and not the static `SimpleBuilder` class as they do not share the same settings. If you don't need dependency injection, using the base library and the static class is recommended.

**Comment by vovanthinh2205** [comment]
I am very grateful to you for helping me and explaining how to fix errors. but now I have a new problem when I use  SimpleSqlBuilder.
I have a request is write an API with 2 parameters as a table and search. 
Api :https://localhost:7202/weatherforecast?table=Product&search=aa
For example, A table Product like
![Capture](https://github.com/mishael-o/Dapper.SimpleSqlBuilder/assets/163615264/358576c9-28d9-4ae1-9ff2-7da6e24ae7d3)
So This is a request for dynamic SQL to follow a condition. With API the example above then it is a result of dynamic SQL 
**`select * from Product where ProductName is like @search and Describe like @search`**.
if is very difficult how to create where in SQL server.so I use table system database SQL server. I take to list column and Type in A table. 
For  example:  
                  ```
var sqlStr = @"SELECT COLUMN_NAME Field, DATA_TYPE Type  FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = @table";
                   var param1 = new DynamicParameters(
                   new
                   {
                       table = table
                   });
                   var tableColumns = dbConnection.Query<dynamic>(sqlStr, param1);
``

![Capture](https://github.com/mishael-o/Dapper.SimpleSqlBuilder/assets/163615264/2a13bd83-6918-4393-a1dc-14ae9d50e250)

it notifies of a error like an image below

![Capture](https://github.com/mishael-o/Dapper.SimpleSqlBuilder/assets/163615264/165791af-c42b-47ce-80b1-631e4e55d301)

**Comment by mishael-o** [comment]
Hi @vovanthinh2205, using `dynamic` with the library for .NET 6 versions and later (we use [InterpolatedStringHandlers](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/interpolated-string-handler)  for perf improvements) is not supported as that is a limitation of `InterpolatedStringHandlers` that are `ref structs`. See the resources below in the dotnet repo where this has been discussed.

- https://github.com/dotnet/roslyn/issues/55609#issuecomment-898636629
- https://github.com/dotnet/roslyn/issues/72606


Also, your approach is not the best, I recommend you do not use `dynamic`. C# has a great type system, so why not take advantage of that. I've included below a better approach to achieve the same result.

```csharp
[HttpGet]
public IEnumerable<Product> GetProducts([FromQuery] string? search = null)
{
    var connectionString = "YOUR_CONNECTION_STRING";

    var builder = _simpleBuilder.CreateFluent()
        .Select($"*")
        .From($"Product");

    using var connection = new SqlConnection(connectionString);

    if (string.IsNullOrWhiteSpace(search))
    {
        return connection.Query<Product>(builder.Sql, builder.Parameters);
    }

    search = $"%{search.Trim()}%";

    builder
        .Where($"ProductName LIKE {search}")
        .OrWhere($"Describe LIKE {search}");

    return connection.Query<Product>(builder.Sql, builder.Parameters);
}

public record Product
{
    public int Id { get; init; }
    public string? ProductName { get; init; }
    public string? Describe { get; init; }
    public DateTime? CreatedDate { get; init; }
}
```

**Comment by vovanthinh2205** [comment]
I fixed  that error with a defined Entity like 

```
public class TableType
    {
         public string Field  { get; set; }
        public string Type { get; set; }
    }
```


Thank you so much for supporting me. I am very grateful and appreciative of you so much. so I am working on my task with the request if is very difficult. For example, I have to do a module Category common with many tables like **(Product, Blog, Category ...)**
There are an API search, an API Create, an API update, an API Delete, and a handle to common for a module. Because I don't define an Entity such as **( Product, Blog,  Category ...)** explicitly like the sample you write above.
I fixed that error ok. but when I use this library of you. I detect a new problem with your core of you. when I use SimpleBuilder.SimpleBuilder.CreateFluent() doesn't understand when I use OrWhere but there is not When. The mean is not  OrWhere alone in syntax leading to an error. Can you improve that feature? I think it's very necessary when people use this library.
The error is the image above.
![Capture](https://github.com/mishael-o/Dapper.SimpleSqlBuilder/assets/163615264/7d3000bf-fc60-496c-b904-908b5d93e61a)

**Comment by mishael-o** [comment]
Hi @vovanthinh2205, good catch, I see what you mean. I will work on enabling the `Where` clauses not to be dependent on each other.

I will release a beta version soon with the update, so you can use that. Will let you in the chat once it is available.

**Comment by vovanthinh2205** [comment]
I am looking forward and excited and waiting for you to release a new version soon. Thank you so much!

**Comment by mishael-o** [comment]
Hi @vovanthinh2205, I have released `beta` version [4.0.0-beta](https://www.nuget.org/packages/Dapper.SimpleSqlBuilder/4.0.0-beta) with the [fixes](https://github.com/mishael-o/Dapper.SimpleSqlBuilder/releases/tag/v4.0.0-beta). You can use that in the meantime until a `main` version is released.

**Comment by vovanthinh2205** [comment]
Thank you so much! I update a new version 4.0.4-beta. I use OrWhere alone in syntax.

**Comment by mishael-o** [comment]
Hi, @vovanthinh2205 that's great. I have released the latest version [4.0.5](https://www.nuget.org/packages/Dapper.SimpleSqlBuilder). You can upgrade to that. 

I will be closing this issue. If you have any other issues to raise you are welcome to create a new issue.
